File: Ruby/Algorithms/quickselect.rb
def quickselect(array,k)
  if array.nil? || array.empty? || k> array.length
    nil
  else
    select(array,0,array.length-1,k)
  end
end

def select(array,left,right,k)
  p_index = partition(array,left,right)
  return array[p_index] if p_index = k-1
  return select(array,p_index+1,right,k) if p_index < k-1
  select(array,left,p_index-1)
end

def partition(array,left,right)
  p_index = left
  pivot = array[right]
  (left...right).each do |i|
    if(array[i] <= pivot)
      array[p_index],array[i] = array[i],array[p_index]
      p_index = p_index+1
    end
  end
  array[right],array[p_index] = array[p_index],array[right]
  p_index
end

puts quickselect [7,5,4,1,2,6,9,8,3],3

File: Java/Algorithms/Quickselect.java
public class Quickselect {
  public static int quickselect (int[] array,int k) {
    if (array.length == 0 || k > array.length) {
      throw new RuntimeException("Illegal Arguments");
    }else {
      return select(array,0,array.length-1,k);
    }
  }
  
  public static int select(int[] array,int left,int right,int k) {
    
    int pIndex = partition(array,left,right);
    if (pIndex == k-1) return array[pIndex];
    if (pIndex <= k-1) return select(array,pIndex+1,right,k);
    return select(array,left,pIndex-1,k);
  }
  

  public static int partition(int[] array, int left, int right) {
    
    int pIndex = left;
    int pivot = array[right];

    for(int i = left; i < right-1; i++) {

      if (array[i] <= pivot) {
        swap(array,i,pIndex);
        pIndex++;
      }

    }
    swap(array,right,pIndex);
    return pIndex;
  }

  public static void swap(int[] array, int i, int j) {
    int tmp = array[j];
    array[j] = array[i];
    array[i] = tmp;
  }

  public static void main(String[] args) {

    System.out.println(quickselect(new int[]{1,2,3,6,5,4,7,8},3));
    
  }

}